samcli/lib/build/workflow_config.py

```diff
@@ -2,60 +2,158 @@
+
+        # When Maven builder exists, add to this list so we can automatically choose a builder based on the supported
+        # manifest
+        "java8": ManifestWorkflowSelector([
+            # Gradle builder needs custom executable paths to find `gradlew` binary
+            JAVA_GRADLE_CONFIG._replace(executable_search_paths=[code_dir, project_dir])
+        ]),
     }
 
-    try:
-        return workflow_config_by_runtime[runtime]
-    except KeyError:
+    if runtime not in selectors_by_runtime:
         raise UnsupportedRuntimeException("'{}' runtime is not supported".format(runtime))
+
+    selector = selectors_by_runtime[runtime]
+
+    try:
+        config = selector.get_config(code_dir, project_dir)
+        return config
+    except ValueError as ex:
+        raise UnsupportedRuntimeException("Unable to find a supported build workflow for runtime '{}'. Reason: {}"
+                                          .format(runtime, str(ex)))
+
+
+class BasicWorkflowSelector(object):
+    """
+    Basic workflow selector that returns the first available configuration in the given list of configurations
+    """
+
+    def __init__(self, configs):
+
+        if not isinstance(configs, list):
+            configs = [configs]
+
+        self.configs = configs
+
+    def get_config(self, code_dir, project_dir):
+        """
+        Returns the first available configuration
+        """
+        return self.configs[0]
+
+
+class ManifestWorkflowSelector(BasicWorkflowSelector):
+    """
+    Selects a workflow by examining the directories for presence of a supported manifest
+    """
+
+    def get_config(self, code_dir, project_dir):
+        """
+        Finds a configuration by looking for a manifest in the given directories.
+
+        Returns
+        -------
+        samcli.lib.build.workflow_config.CONFIG
+            A supported configuration if one is found
+
+        Raises
+        ------
+        ValueError
+            If none of the supported manifests files are found
+        """
+
+        # Search for manifest first in code directory and then in the project directory.
+        # Search order is important here because we want to prefer the manifest present within the code directory over
+        # a manifest present in project directory.
+        search_dirs = [code_dir, project_dir]
+        LOG.debug("Looking for a supported build workflow in following directories: %s", search_dirs)
+
+        for config in self.configs:
+
+            if any([self._has_manifest(config, directory) for directory in search_dirs]):
+                return config
+
+        raise ValueError("None of the supported manifests '{}' were found in the following paths '{}'".format(
+                         [config.manifest_name for config in self.configs],
+                         search_dirs))
+
+    @staticmethod
+    def _has_manifest(config, directory):
+        return os.path.exists(os.path.join(directory, config.manifest_name))
```

Comment:
One day you will use pathlib.. one day..

Reply:
when we move to py3. Pathlib in py2 is really confusing